funderfinder/utils/list_gsoc.py

```diff
@@ -0,0 +1,263 @@
+    links = [
+        extract_listing_link(link_elt) for link_elt in soup.find_all("a", href=True)
+    ]
+    repos = []
+    for link in links:
+        if not link:
+            continue
+        gh_links = get_link_matches(link)
+        repos.extend(gh_links)
+    page_link = get_link_matches(project_page.text)
+    repos.extend(page_link)
+    student_projects = get_early_archive_listing_links(soup)
+    student_project_links = [
+        extract_listing_link(link_container) for link_container in student_projects
+    ]
+    for student_project_link in student_project_links:
+        if not student_project_link:
+            continue
+        student_project_page = requests.get(student_project_link)
+        student_page_links = get_link_matches(student_project_page.text)
+        if student_page_links:
+            repos.extend(student_page_links)
+        time.sleep(SCRAPE_DELAY)
+    return list(set(repos))
+
+
+def get_early_archive_project(container: bs4.BeautifulSoup, year: int) -> dict:
+    """
+    Get a pre-2016 project's metadata
+    :param container: BeautifulSoup container containing the project's name and link to GSOC's detail page
+    :param year: Year the project appeared (a project may appear in more than one year)
+    :return: Dict containing project's name, link to detail page, and any repos/orgs we found for the project
+    """
+    name = container.find("a").text.strip()
+    link = extract_listing_link(container)
+    repos = get_early_archive_repos(link)
+    return {"name": name, "link": link, "repos": repos, "year": year}
+
+
+def get_early_archive_year_projects(link: str) -> iter:
+    """
+    Retrieves all project metadata for a pre-2016 GSOC yearly project listing link
+    :param link: Link to the year's listing page (e.g. https://www.google-melange.com/archive/gsoc/2015)
+    :return: A generator of dicts containing project metadata
+    """
+    listing_page = requests.get(link)
+    soup = bs4.BeautifulSoup(listing_page.text, features="html.parser")
+    project_containers = get_early_archive_listing_links(soup)
+    year = int(link.strip().strip("/").split("/")[-1])
+    for container in project_containers:
+        meta = get_early_archive_project(container, year)
+        yield meta
+        time.sleep(SCRAPE_DELAY)
+
+
+def get_projects_before_2016() -> iter:
+    """
+    Retrieves project metadata from before 2016. GSOC uses an older website format for these
+    :return: A generator of dicts containing project metadata
+    """
+    listing_page = requests.get("https://www.google-melange.com/archive/gsoc")
+    soup = bs4.BeautifulSoup(listing_page.text, features="html.parser")
+    year_link_containers = get_early_archive_listing_links(soup)
+    for container in year_link_containers:
+        year_link = extract_listing_link(container)
+        if not year_link:
+            continue
+        print(f"Getting projects for {year_link}")
```

Comment:
I'm fine with print statements. Perhaps we've already talked about this but in the off-chance we haven't: should this codebase use `print` or a logging statements for information like this? Like I said, I'm fine with print, but if you forsee that using the `logging` module will have benefits, I just wanted to raise it now.

Reply:
Logging is better (and more consistent with what we've done elsewhere), I just wasn't thinking carefully about this. Updated!